Make __owned functional, maybe

huonw opened the pull request:
~~It's hard to tell if this does anything until @gottesmm's work lands, given parameters are all @owned in SIL by default anyway.~~ I found `-enable-guaranteed-normal-arguments`.

huonw committed:
[AST] Explicitly track things marked __owned.

huonw committed:
[SIL] Ensure __owned arguments are @owned.

No matter what the default is, an __owned argument should end up being
@owned in SIL. Currently this is... hard to test, since arguments are
owned by default.

Part of rdar://problem/34668110.

huonw commented:
There's a lot of noise from adding `Owned` to `ValueOwnership` etc. and propagating that around; the interesting bit is in the second commit (in SILFunctionType.cpp). I don't really know what's going on (@gottesmm cribbed it off the shared variant, also without understanding) and it's impossible to test with the defaults the way they are, but it doesn't seem to crash. I'll be trying to understand the code, but it seemed worth submitting for now.

huonw commented:
@swift-ci please smoke test

rjmccall commented on include/swift/ABI/MetadataValues.h:
```diff
@@ -643,7 +643,8 @@ class TargetParameterTypeFlags {
     InOutMask    = 1 << 0,
     SharedMask   = 1 << 1,
-    VariadicMask = 1 << 2,
+    OwnedMask    = 1 << 2,
```
I don't think we need both `shared` and `owned` here, since `owned` is the opposite of `shared`.  But actually, I think you ought to restructure this so that there's a 2-bit ownership field that just stores a ValueOwnership value.

Although giving it 2 bits seems unwisely miserly when we already have two values.   In practice, ParameterTypeFlags is 32/64 bits, right?  We could give it... well, at least 4.

rjmccall commented on lib/SIL/SILFunctionType.cpp:
```diff
@@ -556,6 +556,34 @@ class DestructureInputs {
+    Inputs.push_back(SILParameterInfo(loweredType, convention));
+
+    maybeAddForeignParameters();
+  }
```
Surely there's some sensible way of just parameterizing visitSharedType here.

huonw commented:
@swift-ci please smoke test Linux platform

huonw committed:
[ABI] Store a ValueOwnership directly, not a manual encoding of it.

huonw committed:
Switch a pile of VarDecl::Specifier::Owneds to ::Default.

A lot of these probably should be Owned, but deciding that is work for
another time.

huonw commented:
@swift-ci please test

huonw commented:
(I don't think those failures are related.)

This is a peculiar patch, in that it doesn't work for many cases (many uses of `__owned` parameters incorrectly end up as having an `@lvalue` type, presumably due to treating `__owned` as a `var` or similar, since a lot of these code paths are driven by not-exhaustively-checked `isVar`/`isLet` calls), but I'm still inclined to merge it, while I work on fixing the problems: I think it is enough for the key think that I believe @gottesmm wants of matching the ABI of foreign functions.

huonw commented:
@swift-ci please test

huonw commented:
@swift-ci please smoke test

huonw committed:
[SIL] Ensure __owned arguments are @owned.

No matter what the default is, an __owned argument should end up being
@owned in SIL. Currently this is... hard to test, since arguments are
owned by default.

Part of rdar://problem/34668110.
